### DISCUSSION TITLE
Add a way for `NetworkBehaviour`s to signal an error to the swarm.

### DISCUSSION THREAD

**Discussion by twittner** [libp2p/rust-libp2p#1968] [Ideas]
A `NetworkBehaviour` may enter an error state and can not proceed in an orderly way. The current API contract provides no means to signal such a condition to the swarm which could then act appropriately. Instead of `panic!` we may want to support more fine-grained ways to handle this, e.g. reset the state of the behaviour or terminate the swarm. One possibility would be to add `NetworkBehaviourAction` constructors, e.g. `ResetBehaviour` and `TerminateSwarm`. When receiving a `ResetBehaviour` the swarm would provide the behaviour with the current peer information so it could clear its internal state and  assume the given peers as connected. On `TerminateSwarm` the swarm would itself shutdown operation.

**Comment by tomaka** [comment]
As for `TerminateSwarm`, the only way the `Swarm` could handle that is to return an error to the user, and the user would then have to destroy the `Swarm`. In other words, `Swarm` producing an error would mean that `TerminateSwarm` has been called.

To me, this mechanism would be redundant with `NetworkBehaviourAction::GenerateEvent` being propagated to the user. The only difference would be that `TerminateSwarm` would have a bit more semantics.

**Comment by tomaka** [comment]
As for `ResetBehaviour`, I guess it makes a bit more sense. Ideally I think that we should pass the list of nodes we're connected to by parameter whenever any of the methods of `NetworkBehaviour` is invoked, but that's a change that is way more complicated than it sounds.

However note that some implementations of `NetworkBehaviour` have to stay in sync with some state within their associated `ProtocolsHandlers`, and that state would be lost anyway in case of a "reset behaviour".
